File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/zxart/RemoteCatalog.java
/**
 * @file
 * @brief Remote implementation of catalog
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs.zxart;

import android.net.Uri;
import android.sax.Element;
import android.sax.RootElement;
import android.text.Html;
import android.util.Xml;

import androidx.annotation.Nullable;

import org.xml.sax.SAXException;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Locale;

import app.zxtune.Log;
import app.zxtune.Util;
import app.zxtune.fs.HtmlUtils;
import app.zxtune.fs.http.MultisourceHttpProvider;

public final class RemoteCatalog extends Catalog {

  private static final String TAG = RemoteCatalog.class.getName();

  //no www. prefix!!!
  private static final String SITE = "https://zxart.ee";
  private static final String API = SITE + "/zxtune/language:eng";
  private static final String ACTION_AUTHORS = "/action:authors";
  private static final String ACTION_PARTIES = "/action:parties";
  private static final String ACTION_TRACKS = "/action:tunes";
  private static final String ACTION_TOP = "/action:topTunes";
  private static final String ACTION_SEARCH = "/action:search";
  private static final String LIMIT = "/limit:%d";
  private static final String AUTHOR_ID = "/authorId:%d";
  private static final String PARTY_ID = "/partyId:%d";
  //private static final String TRACK_ID = "/tuneId:%d";
  private static final String ALL_TRACKS_QUERY = API + ACTION_TRACKS;
  private static final String DOWNLOAD_QUERY = SITE + "/file/id:%d";
  private static final String ALL_AUTHORS_QUERY = API + ACTION_AUTHORS;
  private static final String AUTHOR_TRACKS_QUERY = ALL_TRACKS_QUERY + AUTHOR_ID;
  private static final String ALL_PARTIES_QUERY = API + ACTION_PARTIES;
  private static final String PARTY_TRACKS_QUERY = ALL_TRACKS_QUERY + PARTY_ID;
  private static final String TOP_TRACKS_QUERY = API + ACTION_TOP + LIMIT;
  private static final String FIND_TRACKS_QUERY = API + ACTION_SEARCH + "/query:%s";

  private final MultisourceHttpProvider http;

  public RemoteCatalog(MultisourceHttpProvider http) {
    this.http = http;
  }

  @Override
  public void queryAuthors(AuthorsVisitor visitor) throws IOException {
    Log.d(TAG, "queryAuthors()");
    final InputStream stream = http.getInputStream(Uri.parse(ALL_AUTHORS_QUERY));
    final RootElement root = createAuthorsParserRoot(visitor);
    performQuery(stream, root);
  }

  @Override
  public void queryAuthorTracks(Author author, TracksVisitor visitor) throws IOException {
    Log.d(TAG, "queryAuthorTracks(author=%d)", author.getId());
    queryTracks(visitor, String.format(Locale.US, AUTHOR_TRACKS_QUERY, author.getId()));
  }

  @Override
  public void queryParties(PartiesVisitor visitor) throws IOException {
    Log.d(TAG, "queryParties()");
    final InputStream stream = http.getInputStream(Uri.parse(ALL_PARTIES_QUERY));
    final RootElement root = createPartiesParserRoot(visitor);
    performQuery(stream, root);
  }

  @Override
  public void queryPartyTracks(Party party, TracksVisitor visitor) throws IOException {
    Log.d(TAG, "queryPartyTracks(party=%d)", party.id);
    queryTracks(visitor, String.format(Locale.US, PARTY_TRACKS_QUERY, party.id));
  }

  @Override
  public void queryTopTracks(int limit, TracksVisitor visitor) throws IOException {
    Log.d(TAG, "queryTopTracks()");
    queryTracks(visitor, String.format(Locale.US, TOP_TRACKS_QUERY, limit));
  }

  private void queryTracks(TracksVisitor visitor, String query) throws IOException {
    final InputStream stream = http.getInputStream(Uri.parse(query));
    final RootElement root = createModulesParserRoot(visitor);
    performQuery(stream, root);
  }

  final boolean searchSupported() {
    return http.hasConnection();
  }

  @Override
  public void findTracks(String query, FoundTracksVisitor visitor) throws IOException {
    Log.d(TAG, "findTracks(query=%s)", query);
    final String uri = String.format(Locale.US, FIND_TRACKS_QUERY, Uri.encode(query));
    final InputStream stream = http.getInputStream(Uri.parse(uri));
    final RootElement root = createModulesParserRoot(visitor);
    performQuery(stream, root);
  }

  public static Uri[] getTrackUris(int id) {
    final String query = String.format(Locale.US, DOWNLOAD_QUERY, id);
    return new Uri[]{Uri.parse(query)};
  }

  private void performQuery(InputStream httpStream, RootElement root)
          throws IOException {
    try {
      final InputStream stream = new BufferedInputStream(httpStream);
      Xml.parse(stream, Xml.Encoding.UTF_8, root.getContentHandler());
    } catch (SAXException e) {
      throw new IOException(e);
    } finally {
      httpStream.close();
    }
  }

  private static RootElement createAuthorsParserRoot(final AuthorsVisitor visitor) {
    final AuthorBuilder builder = new AuthorBuilder();
    final RootElement result = createRootElement();
    final Element data = result.getChild("responseData");
    data.getChild("totalAmount").setEndTextElementListener(body -> {
      final Integer count = HtmlUtils.tryGetInteger(body);
      if (count != null) {
        visitor.setCountHint(count);
      }
    });
    final Element item = data.getChild("authors").getChild("author");
    item.setEndElementListener(() -> {
      final Author res = builder.captureResult();
      if (res != null) {
        visitor.accept(res);
      }
    });
    item.getChild("id").setEndTextElementListener(builder::setId);
    item.getChild("title").setEndTextElementListener(builder::setNickname);
    item.getChild("realName").setEndTextElementListener(builder::setName);
    return result;
  }

  private static class AuthorBuilder {

    @Nullable
    private Integer id;
    @Nullable
    private String nickname;
    @Nullable
    private String name;

    final void setId(String val) {
      id = Integer.valueOf(val);
    }

    final void setNickname(String val) {
      nickname = val;
    }

    final void setName(String val) {
      name = val;
    }

    @Nullable
    final Author captureResult() {
      final Author res = isValid() ? new Author(id, nickname, name != null ? name : "") : null;
      id = null;
      nickname = null;
      name = null;
      return res;
    }

    private boolean isValid() {
      return id != null && nickname != null;
    }
  }

  private static RootElement createPartiesParserRoot(final PartiesVisitor visitor) {
    final PartiesBuilder builder = new PartiesBuilder();
    final RootElement result = createRootElement();
    final Element data = result.getChild("responseData");
    data.getChild("totalAmount").setEndTextElementListener(body -> {
      final Integer count = HtmlUtils.tryGetInteger(body);
      if (count != null) {
        visitor.setCountHint(count);
      }
    });
    final Element item = data.getChild("parties").getChild("party");
    item.setEndElementListener(() -> {
      final Party res = builder.captureResult();
      if (res != null) {
        visitor.accept(res);
      }
    });
    item.getChild("id").setEndTextElementListener(builder::setId);
    item.getChild("title").setEndTextElementListener(builder::setName);
    item.getChild("year").setEndTextElementListener(builder::setYear);
    return result;
  }

  private static class PartiesBuilder {

    @Nullable
    private Integer id;
    @Nullable
    private String name;
    private int year;

    final void setId(String val) {
      id = Integer.valueOf(val);
    }

    final void setName(String val) {
      name = val;
    }

    final void setYear(String val) {
      try {
        year = Integer.valueOf(val);
      } catch (NumberFormatException e) {
        year = 0;
      }
    }

    @Nullable
    final Party captureResult() {
      final Party res = isValid() ? new Party(id, name, year) : null;
      id = null;
      name = null;
      year = 0;
      return res;
    }

    private boolean isValid() {
      return id != null && name != null;
    }
  }

  private static RootElement createModulesParserRoot(final TracksVisitor visitor) {
    final ModuleBuilder builder = new ModuleBuilder();
    final RootElement result = createRootElement();
    final Element data = result.getChild("responseData");
    data.getChild("totalAmount").setEndTextElementListener(body -> {
      final Integer count = HtmlUtils.tryGetInteger(body);
      if (count != null) {
        visitor.setCountHint(count);
      }
    });
    final Element item = data.getChild("tunes").getChild("tune");
    item.setEndElementListener(() -> {
      final Track result1 = builder.captureResult();
      if (result1 != null) {
        visitor.accept(result1);
      }
    });
    bindXmlActions(item, builder);
    return result;
  }

  //TODO: remove C&P
  private static RootElement createModulesParserRoot(final FoundTracksVisitor visitor) {
    final ModuleBuilder builder = new ModuleBuilder();
    final RootElement result = createRootElement();
    final Element data = result.getChild("responseData");
    data.getChild("totalAmount").setEndTextElementListener(body -> {
      final Integer count = HtmlUtils.tryGetInteger(body);
      if (count != null) {
        visitor.setCountHint(count);
      }
    });
    final Element item = data.getChild("tunes").getChild("tune");
    item.setEndElementListener(() -> {
      final Author author = builder.captureResultAuthor();
      final Track result1 = builder.captureResult();
      if (result1 != null && author != null) {
        visitor.accept(author, result1);
      }
    });
    bindXmlActions(item, builder);
    return result;
  }

  private static void bindXmlActions(Element item, final ModuleBuilder builder) {
    item.getChild("id").setEndTextElementListener(builder::setId);
    //CDATA
    item.getChild("originalFileName").setEndTextElementListener(body -> builder.setFilename(Html.fromHtml(body).toString()));
    item.getChild("title").setEndTextElementListener(builder::setTitle);
    item.getChild("internalAuthor").setEndTextElementListener(builder::setInternalAuthor);
    item.getChild("internalTitle").setEndTextElementListener(builder::setInternalTitle);
    item.getChild("votes").setEndTextElementListener(builder::setVotes);
    item.getChild("time").setEndTextElementListener(builder::setDuration);
    item.getChild("year").setEndTextElementListener(builder::setYear);
    item.getChild("compo").setEndTextElementListener(builder::setCompo);
    item.getChild("partyplace").setEndTextElementListener(builder::setPartyplace);
    item.getChild("authors").getChild("id").setEndTextElementListener(builder::setAuthorId);
  }

  private static class ModuleBuilder {

    @Nullable
    private Integer id;
    @Nullable
    private String filename;
    private String title = "";
    private String internalAuthor = "";
    private String internalTitle = "";
    private String votes = "";
    private String duration = "";
    private int year;
    @Nullable
    private String compo;
    private int partyplace;
    @Nullable
    private Integer authorId;

    ModuleBuilder() {
      reset();
    }

    final void setId(String val) {
      id = Integer.valueOf(val);
    }

    final void setFilename(String val) {
      filename = val.trim();
      if (filename.length() == 0) {
        filename = "unknown";
      }
    }

    final void setTitle(String val) {
      title = val;
    }

    final void setInternalAuthor(String val) {
      internalAuthor = val;
    }

    final void setInternalTitle(String val) {
      internalTitle = val;
    }

    final void setVotes(String val) {
      votes = val;
    }

    final void setDuration(String val) {
      duration = val;
    }

    final void setYear(String val) {
      try {
        year = Integer.valueOf(val);
      } catch (NumberFormatException e) {
        year = 0;
      }
    }

    final void setCompo(String val) {
      compo = val;
    }

    final void setPartyplace(String val) {
      try {
        partyplace = Integer.valueOf(val);
      } catch (NumberFormatException e) {
        partyplace = 0;
      }
    }

    final void setAuthorId(String val) {
      try {
        authorId = Integer.valueOf(val);
      } catch (NumberFormatException e) {
        authorId = null;
      }
    }

    @Nullable
    final Author captureResultAuthor() {
      if (authorId != null) {
        final String fakeName = "Author" + authorId;
        final Author res = new Author(authorId, fakeName, "");
        authorId = null;
        return res;
      } else {
        return null;
      }
    }

    @Nullable
    final Track captureResult() {
      title = Util.formatTrackTitle(internalAuthor, internalTitle, title);
      final Track res = isValid()
              ? new Track(id, filename, title, votes, duration, year, compo, partyplace)
              : null;
      reset();
      return res;
    }

    private void reset() {
      id = null;
      filename = null;
      year = 0;
      partyplace = 0;
      votes = "";
      duration = "";
      title = "";
      internalAuthor = "";
      internalTitle = "";
      compo = null;
      authorId = null;
    }

    private boolean isValid() {
      return id != null && filename != null;
    }
  }

  private static RootElement createRootElement() {
    return new RootElement("response");
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/zxart/RemoteCatalog.kt
/**
 * @file
 * @brief Remote implementation of catalog
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.fs.zxart

import android.net.Uri
import android.sax.Element
import android.sax.RootElement
import android.text.Html
import app.zxtune.Logger
import app.zxtune.Util
import app.zxtune.fs.HtmlUtils.tryGetInteger
import app.zxtune.fs.http.MultisourceHttpProvider
import app.zxtune.utils.Xml
import app.zxtune.utils.ifNotNulls
import java.io.BufferedInputStream

private val LOG = Logger(RemoteCatalog::class.java.name)

open class RemoteCatalog(private val http: MultisourceHttpProvider) : Catalog {

    override fun queryAuthors(visitor: Catalog.Visitor<Author>) {
        LOG.d { "queryAuthors()" }
        val root = createAuthorsParserRoot(visitor)
        performQuery(Uris.forAuthors(), root)
    }

    override fun queryAuthorTracks(author: Author, visitor: Catalog.Visitor<Track>) {
        LOG.d { "queryAuthorTracks(author=${author.id})" }
        queryTracks(visitor, Uris.forTracksOf(author))
    }

    override fun queryParties(visitor: Catalog.Visitor<Party>) {
        LOG.d { "queryParties()" }
        val root = createPartiesParserRoot(visitor)
        performQuery(Uris.forParties(), root)
    }

    override fun queryPartyTracks(party: Party, visitor: Catalog.Visitor<Track>) {
        LOG.d { "queryPartyTracks(party=${party.id}" }
        queryTracks(visitor, Uris.forTracksOf(party))
    }

    override fun queryTopTracks(limit: Int, visitor: Catalog.Visitor<Track>) {
        LOG.d { "queryTopTracks()" }
        queryTracks(visitor, Uris.forTop(limit))
    }

    private fun queryTracks(visitor: Catalog.Visitor<Track>, uri: Uri) {
        val root = createModulesParserRoot(visitor)
        performQuery(uri, root)
    }

    open fun searchSupported() = http.hasConnection()

    override fun findTracks(query: String, visitor: Catalog.FoundTracksVisitor) {
        LOG.d { "findTracks(query=${query})" }
        val root = createModulesParserRoot(visitor)
        performQuery(Uris.forSearch(query), root)
    }

    private fun performQuery(uri: Uri, root: RootElement) =
        http.getInputStream(uri).let { httpStream ->
            Xml.parse(BufferedInputStream(httpStream), root.contentHandler)
        }

    companion object {
        @JvmStatic
        fun getTrackUris(id: Int) = arrayOf(Uris.forDownload(id))
    }
}

private class Uris {
    private val delegate = Uri.Builder()

    init {
        delegate.scheme("https").authority("zxart.ee") // no www. prefix!
    }

    fun action(action: String) = apply {
        delegate.appendPath("zxtune").appendEncodedPath("language:eng")
            .appendEncodedPath("action:${action}")
    }

    fun field(type: String, id: Int) = apply {
        delegate.appendEncodedPath("${type}:${id}")
    }

    fun field(type: String, value: String) = apply {
        delegate.appendEncodedPath("${type}:${Uri.encode(value)}")
    }

    fun download(id: Int) = run {
        delegate.appendPath("file")
        field("id", id)
    }

    fun build(): Uri = delegate.build()

    companion object {
        fun forAuthors() = Uris().action("authors").build()

        fun forTracksOf(author: Author) = forTracksOf("authorId", author.id)

        fun forParties() = Uris().action("parties").build()

        fun forTracksOf(party: Party) = forTracksOf("partyId", party.id)

        fun forTop(limit: Int) = Uris().action("topTunes").field("limit", limit).build()

        fun forSearch(query: String) = Uris().action("search").field("query", query).build()

        private fun forTracksOf(scope: String, id: Int) =
            Uris().action("tunes").field(scope, id).build()

        fun forDownload(id: Int) = Uris().download(id).build()
    }
}

private class AuthorBuilder {
    private var id: Int? = null
    private var nickname: String? = null
    private var name: String? = null

    fun setId(value: String?) {
        id = tryGetInteger(value)
    }

    fun setNickname(value: String?) {
        nickname = value
    }

    fun setName(value: String?) {
        name = value
    }

    fun captureResult() = ifNotNulls(id, nickname, name.orEmpty(), ::Author).also {
        id = null
        nickname = null
        name = null
    }
}

private class PartiesBuilder {
    private var id: Int? = null
    private var name: String? = null
    private var year = 0

    fun setId(value: String?) {
        id = tryGetInteger(value)
    }

    fun setName(value: String?) {
        name = value
    }

    fun setYear(value: String?) {
        year = tryGetInteger(value) ?: 0
    }

    fun captureResult() = ifNotNulls(id, name, year, ::Party).also {
        id = null
        name = null
        year = 0
    }
}

private class ModuleBuilder {
    private var id: Int? = null
    private var filename: String? = null
    private var title = ""
    private var internalAuthor = ""
    private var internalTitle = ""
    private var votes = ""
    private var duration = ""
    private var year = 0
    private var compo: String = ""
    private var partyplace = 0
    private var authorId: Int? = null

    init {
        reset()
    }

    fun setId(value: String?) {
        id = value?.toIntOrNull()
    }

    fun setFilename(value: String?) {
        filename = value?.trim()?.takeIf { it.isNotEmpty() }?.let { Html.fromHtml(it).toString() }
            ?: "unknown"
    }

    fun setTitle(value: String?) {
        title = value.orEmpty()
    }

    fun setInternalAuthor(value: String?) {
        internalAuthor = value.orEmpty()
    }

    fun setInternalTitle(value: String?) {
        internalTitle = value.orEmpty()
    }

    fun setVotes(value: String?) {
        votes = value.orEmpty()
    }

    fun setDuration(value: String?) {
        duration = value.orEmpty()
    }

    fun setYear(value: String?) {
        year = tryGetInteger(value) ?: 0
    }

    fun setCompo(value: String?) {
        compo = value.orEmpty()
    }

    fun setPartyplace(value: String?) {
        partyplace = tryGetInteger(value) ?: 0
    }

    fun setAuthorId(value: String?) {
        authorId = tryGetInteger(value)
    }

    fun captureResultAuthor() = ifNotNulls(authorId, "Author$authorId", "", ::Author).also {
        authorId = null
    }

    fun captureResult() = ifNotNulls(id, filename) { id, filename ->
        Track(
            id,
            filename,
            Util.formatTrackTitle(internalAuthor, internalTitle, title),
            votes,
            duration,
            year,
            compo,
            partyplace
        )
    }.also {
        reset()
    }

    private fun reset() {
        id = null
        filename = null
        year = 0
        partyplace = 0
        votes = ""
        duration = ""
        title = ""
        internalAuthor = ""
        internalTitle = ""
        compo = ""
        authorId = null
    }
}

private fun createAuthorsParserRoot(visitor: Catalog.Visitor<Author>) = createRootElement().apply {
    getChild("responseData").run {
        getChild("authors").getChild("author").run {
            val builder = AuthorBuilder()
            setEndElementListener {
                builder.captureResult()?.let { res ->
                    visitor.accept(res)
                }
            }
            getChild("id").setEndTextElementListener(builder::setId)
            getChild("title").setEndTextElementListener(builder::setNickname)
            getChild("realName").setEndTextElementListener(builder::setName)
        }
    }
}

private fun createPartiesParserRoot(visitor: Catalog.Visitor<Party>) = createRootElement().apply {
    getChild("responseData").run {
        getChild("parties").getChild("party").run {
            val builder = PartiesBuilder()
            setEndElementListener {
                builder.captureResult()?.let { res ->
                    visitor.accept(res)
                }
            }
            getChild("id").setEndTextElementListener(builder::setId)
            getChild("title").setEndTextElementListener(builder::setName)
            getChild("year").setEndTextElementListener(builder::setYear)
        }
    }
}

private fun createModulesParserRoot(visitor: Catalog.Visitor<Track>) =
    createModulesParserRoot { builder ->
        builder.captureResult()?.let {
            visitor.accept(it)
        }
    }

private fun createModulesParserRoot(visitor: Catalog.FoundTracksVisitor) =
    createModulesParserRoot { builder ->
        builder.captureResultAuthor()?.let { author ->
            builder.captureResult()?.let { track ->
                visitor.accept(author, track)
            }
        }
    }

private fun createModulesParserRoot(
    callback: (ModuleBuilder) -> Unit
) = createRootElement().apply {
    getChild("responseData").run {
        getChild("tunes").getChild("tune").run {
            val builder = ModuleBuilder()
            setEndElementListener {
                callback(builder)
            }
            bindXmlActions(this, builder)
        }
    }
}

private fun bindXmlActions(item: Element, builder: ModuleBuilder) = with(item) {
    getChild("id").setEndTextElementListener(builder::setId)
    //CDATA
    getChild("originalFileName").setEndTextElementListener(builder::setFilename)
    getChild("title").setEndTextElementListener(builder::setTitle)
    getChild("internalAuthor").setEndTextElementListener(builder::setInternalAuthor)
    getChild("internalTitle").setEndTextElementListener(builder::setInternalTitle)
    getChild("votes").setEndTextElementListener(builder::setVotes)
    getChild("time").setEndTextElementListener(builder::setDuration)
    getChild("year").setEndTextElementListener(builder::setYear)
    getChild("compo").setEndTextElementListener(builder::setCompo)
    getChild("partyplace").setEndTextElementListener(builder::setPartyplace)
    getChild("authors").getChild("id").setEndTextElementListener(builder::setAuthorId)
}

private fun createRootElement() = RootElement("response")
